addons/controllers/clusterbootstrap_controller.go

```diff
@@ -500,19 +500,23 @@ func (r *ClusterBootstrapReconciler) mergeClusterBootstrapPackagesWithTemplate(
 	//    4. All packages, including additional packages, can't be deleted (meaning the package refName can't be changed, only allow version bump)
 	//    5. We will keep users' customization on valuesFrom of each package, users are responsible for the correctness of the content they put in will work with the next version.
 	packages := make([]*runtanzuv1alpha3.ClusterBootstrapPackage, 0)
-	if updatedClusterBootstrap.Spec.CNI == nil {
-		log.Info("no CNI package specified in ClusterBootstrap, should not happen. Continue with CNI in ClusterBootstrapTemplate of new TKR")
-		updatedClusterBootstrap.Spec.CNI = clusterBootstrapTemplate.Spec.CNI.DeepCopy()
-	} else {
-		// We don't allow change to the CNI selection once it starts running, however we allow version bump
-		// TODO: check correctness of the following statement, as we still allow version bump
-		// ClusterBootstrap webhook will make sure the package RefName always match the original CNI
-		updatedCNI, cniNamePrefix, err := util.GetBootstrapPackageNameFromTKR(r.context, r.Client, updatedClusterBootstrap.Spec.CNI.RefName, cluster)
-		if err != nil {
-			errorMsg := fmt.Sprintf("unable to find any CNI bootstrap package prefixed with '%s' for ClusterBootstrap %s/%s in TKR", cniNamePrefix, cluster.Namespace, cluster.Name)
-			return nil, errors.Wrap(err, errorMsg)
+
+	_, usingCustomCNIProvider := updatedClusterBootstrap.Annotations[constants.UseCustomCNIProvider]
+	if !(usingCustomCNIProvider && updatedClusterBootstrap.Spec.CNI == nil) {
```

Comment:
The requirements to indicate that the user wants to use a custom cni is to annotate the clusterbootstrap *and* that Spec.CNI == nil. 

So the opposite (using a built in provider like antrea or calico) is the oposite of that: 
  
!(usingCustomCNIProvider && updatedClusterBootstrap.Spec.CNI == nil) = usingBuiltInProvider (new variable I will introduce to make clear). 

The inside check (updatedClusterBootstrap.Spec.CNI == nil ) is done because there is a chance that usingCustomCNIProvider = false && Spec.CNI== nil, in which case we will enter the loop. 

Will reformat code to make this clearer.

Reply:
done

Reply:
done